ui/imports/shared/popups/keycard/states/EnterName.qml

```diff
@@ -0,0 +1,110 @@
+import QtQuick 2.14
+import QtQuick.Layouts 1.14
+import QtQuick.Controls 2.14
+
+import StatusQ.Core 0.1
+import StatusQ.Core.Theme 0.1
+import StatusQ.Controls 0.1
+
+import utils 1.0
+
+import "../helpers"
+
+Item {
+    id: root
+
+    property var sharedKeycardModule
+
+    signal validation(bool result)
+
+    QtObject {
+        id: d
+        property bool entryValid: false
+
+        function updateValidity() {
+            d.entryValid = keycardName.text.trim().length > 0 && keycardName.text !== root.sharedKeycardModule.keyPairStoredOnKeycard.name
+            root.validation(d.entryValid)
+        }
+    }
+
+    Component.onCompleted: {
+        keycardName.input.edit.forceActiveFocus()
+    }
+
+    ColumnLayout {
+        anchors.fill: parent
+        anchors.topMargin: Style.current.xlPadding
+        anchors.bottomMargin: Style.current.halfPadding
+        anchors.leftMargin: Style.current.xlPadding
+        anchors.rightMargin: Style.current.xlPadding
+        spacing: Style.current.padding
+
+        StatusBaseText {
+            id: title
+            Layout.alignment: Qt.AlignCenter
+            font.weight: Font.Bold
+        }
+
+        StatusInput {
+            id: keycardName
+            Layout.preferredWidth: Constants.keycard.general.keycardNameInputWidth
+            Layout.alignment: Qt.AlignCenter
+            charLimit: Constants.keycard.general.keycardNameLength
+            placeholderText: qsTr("Keycard name")
+            text: root.sharedKeycardModule.keyPairStoredOnKeycard.name
+            input.acceptReturn: true
+
+            onTextChanged: {
+                d.updateValidity()
+                if (d.entryValid) {
+                    root.sharedKeycardModule.setKeycarName(text)
+                }
+            }
+
+            onKeyPressed: {
+                if (d.entryValid &&
+                        (input.edit.keyEvent === Qt.Key_Return ||
+                         input.edit.keyEvent === Qt.Key_Enter)) {
+                    event.accepted = true
+                    root.sharedKeycardModule.currentState.doPrimaryAction()
```

Comment:
@stefandunca as you can see primary action will be called if user presses enter key while the status input field has focus, that is the same action which will be done as if user would click the most right button on the popup when this state is displayed. General rule is that if the back button is displayed then click on that button calls back action for the displayed state, if cancel button is displayed it always does tertiary action, if there is a single button, but not cancel then primary action is called, it there are two buttons but non of them is cancel button then the most right button calls primary action and button next to it calls secondary action. With this approach we have kind of general solution which covers all flows, otherwise we would have a lot of repeated code.

Reply:
How many states can you be in using this component then? How can I see/know this list when debugging?

> With this approach we have kind of general solution which covers all flows, otherwise we would have a lot of repeated code.

The general FSM solution is that you set up all the states in one place clearly. Then you set up all the transitions between states and their specific events by friendly names, without repeating a single line of code, then in the QMLs and Internal events, you just call the SM events by friendly names. Don't even have to duplicate a single check because the state machine is already configured and pre setup.

With an FSM approach when you debug you just track the current state and the events that were triggered and you know the next state.

Reply:
> How many states can you be in using this component then? How can I see/know this list when debugging?

You need to follow appropriate figma file and code.